src/unix/linux-iouring.c

```diff
@@ -0,0 +1,310 @@
+                       UV__IORING_OFF_SQES);
+
+  if (ring->sq.sqes == MAP_FAILED) {
+    status = errno;
+    munmap_checked(ring->sq.head, ring->sq.map_sz);
+    uv__close(fd);
+    return UV__ERR(status);
+  }
+
+  /* Map CQ */
+  ring->cq.map_sz = params.cq_off.cqes + params.cq_entries *
+      sizeof(*ring->cq.cqes);
+  cq_ptr = mmap(NULL,
+                ring->cq.map_sz,
+                PROT_READ | PROT_WRITE,
+                MAP_SHARED | MAP_POPULATE,
+                fd,
+                UV__IORING_OFF_CQ_RING);
+
+  if (cq_ptr == MAP_FAILED) {
+    status = errno;
+    munmap_checked(ring->sq.sqes, sqe_map_size);
+    munmap_checked(ring->sq.head, ring->sq.map_sz);
+    uv__close(fd);
+    return UV__ERR(status);
+  }
+
+  ring->cq.head = (unsigned*)(cq_ptr + params.cq_off.head);
+  ring->cq.tail = (unsigned*)(cq_ptr + params.cq_off.tail);
+  ring->cq.ring_mask = (unsigned*)(cq_ptr + params.cq_off.ring_mask);
+  ring->cq.ring_entries = (unsigned*)(cq_ptr + params.cq_off.ring_entries);
+  ring->cq.overflow = (unsigned*)(cq_ptr + params.cq_off.overflow);
+  ring->cq.cqes = (struct uv__io_uring_cqe*)(cq_ptr + params.cq_off.cqes);
+
+  ring->pending = 0;
+
+  return 0;
+}
+
+
+void uv__io_uring_destroy(struct uv__io_uring* ring) {
+  munmap_checked(ring->sq.sqes,
+                 *ring->sq.ring_entries * sizeof(*ring->sq.sqes));
+  munmap_checked(ring->sq.head, ring->sq.map_sz);
+  munmap_checked(ring->cq.head, ring->cq.map_sz);
+  uv__close(ring->ring_fd);
+}
+
+struct uv__io_uring_sqe* uv__io_uring_get_sqe(struct uv__io_uring* ring) {
+  struct uv__io_uring_sqe* sqe;
+  unsigned idx;
+
+  /* Barrier before reading the SQ head. */
+  mem_barrier();
+  if (*ring->sq.tail - *ring->sq.head == *ring->sq.ring_entries) {
+    /* SQ is full */
+    return NULL;
+  }
+
+  idx = *ring->sq.tail & *ring->sq.ring_mask;
+  sqe = &ring->sq.sqes[idx];
+
+  memset(sqe, 0, sizeof(*sqe));
+
+  return sqe;
+}
+
+
+int uv__io_uring_submit(struct uv__io_uring* ring) {
+  unsigned idx;
+  int rc;
+
+  idx = *ring->sq.tail & *ring->sq.ring_mask;
+  ring->sq.array[idx] = idx;
+  /* Ensure SQE stores are visible before we update the tail. */
+  mem_barrier();
+
+  (*ring->sq.tail)++;
+  /* Ensure kernel sees tail update. */
+  mem_barrier();
+  rc = uv__io_uring_enter(ring->ring_fd, 1, 0, 0, NULL);
+
+  if (rc == -1) {
+    /* The kernel doesn't increment the head on failure. */
+    (*ring->sq.tail)--;
+    mem_barrier();
+  }
+
+  return rc;
+}
+
+
+void uv__io_uring_get_cqe(struct uv__io_uring* ring,
+                          struct uv__io_uring_cqe** cqe_ptr) {
+  unsigned head;
+
+  head = *ring->cq.head;
+  /* Barrier before reading the CQ tail. */
+  mem_barrier();
+  if (head != *ring->cq.tail)
+    *cqe_ptr = &ring->cq.cqes[head & *ring->cq.ring_mask];
+  else
+    *cqe_ptr = NULL;
+}
+
+
+/* TODO I'd rather this be static inline, but atomic-ops.h (mem_barrier)
+ * includes internal.h includes linux-iouring.h because it needs uv__io_uring.
+ */
```

Comment:
I wanted the function definition in the header so it can be inlined in linux-core.c#198. Maybe I'm missing it but I don't see a way to do that without moving `mem_barrier` outside of atomic-ops.h.

(Maybe LTO makes the concern moot anyway.)

(PS: I also realized the `inline` specifier is never used in libuv. I've removed it in the latest revision.)